module/zfs/dmu_recv.c

```diff
@@ -1928,9 +1930,12 @@ receive_read_record(struct receive_arg *ra)
 	{
 		struct drr_object *drro = &ra->rrd->header.drr_u.drr_object;
 		uint32_t size = DRR_OBJECT_PAYLOAD_SIZE(drro);
-		void *buf = kmem_zalloc(size, KM_SLEEP);
+		void *buf = NULL;
 		dmu_object_info_t doi;
 
+		if (size != 0)
+			buf = kmem_zalloc(size, KM_SLEEP);
+
 		err = receive_read_payload_and_next_header(ra, size, buf);
 		if (err != 0) {
 			kmem_free(buf, size);
```

Comment:
We should wrap this `kmem_free(buf, size);` with a `if (size != 0)` check too to.  If we don't its possible the spl per-allocation tracking code will get confused when we try and free something we never allocated.

Reply:
I don't think this is necessary. It is standard practice in c for `free()` functions to take no action if provided a `NULL` pointer, and the kmem code handles that correctly. Additionally, if we did want to do this there are 3 other places in the receive code where we would also want to add this check, which seems a bit unnecessary.

Reply:
Indeed it does, alright.